### PR TITLE
LPS-107404 and LPS-107195 Changes to HttpImpl

### PR DESCRIPTION
**Pull request by csierra**
/cc @arthurchan35

to use URI class. I discussed this with Brian and he said he remembered we did not use URI class mainly for performance reasons. We are using it in some places already and there are known problems with our current impl. In this pull we switch to use URI class and cache it in a thread local. 

There are two issues in this PR... one that update HttpImpl to use URI class and other that makes escapeRedirect to use HttpImpl.

Let us know if this would be ok so we can send it to Brian. 

Thx.

Carlos.

### CODE REVIEW

**Comment by csierra** [comment]
ci:test:sf

**Comment by csierra** [comment]
ci:test:relevant

**Comment by shuyangzhou** [comment]
This could cause some performance regression.
@slnn please run a regression test first.

**Comment by csierra** [comment]
yes... thanks for testing it. 

Carlos.

**Comment by slnn** [comment]
OK, got it and will update test results when I got them

**Comment by slnn** [comment]
Hi @shuyangzhou 

Here are test results:

Login:

Portal Version | Git Id | Session Count | Mean Time of Login | Mean Time of Logout | Session Count Change （%）
-- | -- | -- | -- | -- | --
BaseLine | b8495cee | 44632 | 265 | 78.5 | -
PR | - | 44170 | 267 | 84.6 | -1.035131744
PR | - | 44209 | 286 | 79.6 | -0.9477504929

WebContent:

Portal Version | Git Id | Session Count | Mean Time of Login | Mean Time of View Page | Session Count Change （%）
-- | -- | -- | -- | -- | --
BaseLine | b8495cee | 56206 | 503 | 151 | -
PR | - | 55227 | 1200 | 348 | -1.741806925
PR | - | 55276 | 604 | 185 | -1.65462762

MessageBoard:

Portal Version | Git Id | Session Count | Mean Time of Login | Mean Time of Add MB Thread | Session Count Change （%）
-- | -- | -- | -- | -- | --
BaseLine | b8495cee | 17794 | 27.7 | 76.9 | -
PR | - | 17772 | 107 | 147 | -0.1236371811

Blog:

Portal Version | Git Id | Session Count | Mean Time of Login | Mean Time of Add Blog Entry | Session Count Change （%）
-- | -- | -- | -- | -- | --
BaseLine | b8495cee | 14712 | 357 | 371 | -
PR | - | 14889 | 527 | 457 | 1.203099511

DocumentLibrary:

Portal Version | Git Id | Session Count | Mean Time of Login | Mean Time of Add DL File | Session Count Change （%）
-- | -- | -- | -- | -- | --
BaseLine | b8495cee | 25512 | 210 | 305 | -
PR | - | 26139 | 239 | 302 | 2.45766698

Wiki:

Portal Version | Git Id | Session Count | Mean Time of Login | Mean Time of Add Wiki Page | Session Count Change （%）
-- | -- | -- | -- | -- | --
BaseLine | b8495cee | 15177 | 243 | 461 | -
PR | - | 15277 | 230 | 452 | 0.6588917441

AssetPublisher:

Portal Version | Git Id | Session Count | Mean Time of Login | Mean Time of View Page with Filter | Session Count Change （%）
-- | -- | -- | -- | -- | --
BaseLine | b8495cee | 25726 | 430 | 177 | -
PR | - | 26211 | 429 | 175 | 1.885252274

Regards
Lily

File: portal-impl/src/com/liferay/portal/util/HttpImpl.java [hunk 1]
```diff
@@ -1899,6 +1901,20 @@ public void close() throws IOException {
...
 
+	private URI _getURI(String url) throws URISyntaxException {
+		Map<String, URI> uriMap = _uriMap.get();
+
+		URI uri = uriMap.get(url);
+
+		if (uri == null) {
+			uri = new URI(url);
+
+			uriMap.put(url, new URI(url));
```
Comment (shuyangzhou) [inline portal-impl/src/com/liferay/portal/util/HttpImpl.java]: @csierra you have a double creation for the URI here, please fix it and rerun the performance test, the regression should get smaller after the fix.

**Comment by shuyangzhou** [comment]
@csierra please see my comment above, fix it and have @slnn rerun the regression test. Thanks!

**Comment by arthurchan35** [comment]
See https://github.com/csierra/liferay-portal/pull/617

**Comment by csierra** [comment]
dang! :man_facepalming: thx for caching that!

**Pull request closed without merging**